cloudferrylib/os/identity/keystone.py

```diff
@@ -609,9 +609,14 @@ def _get_user_tenants_roles_by_db(self, tenant_list, user_list):
                 continue
 
             roles_ids = ast.literal_eval(roles_field)['roles']
+            if isinstance(roles_ids[0], dict):
```

Comment:
Verifying object type at runtime for these types of failures looks like a dirty hack. Also you try to retrieve first element of object (`roles_ids[0]`), which may result in `IndexError` for empty list.